README.md
# fatint

Dart library for arbitrary sized integer values. Dart VM by default 
supports arbitrary sized integers but it is not the case when compiled
to Javascript on the browser. This package provides a common big number
interface for DartVM and browser and provides respective implementations.

# Simple examples

```dart
BigInteger i1 = bigInt(25) % bigInt(7); // => 4
BigInteger i2 = bigHex('abcd1234') * bigHex('beef'); // => 0x802297f6968c
```

# Advantage

+ Clean API complying to to Dart's int
+ Efficient equations using `BigIntegerRef` and `zero`

CHANGELOG.md
# Changelog

## 2.0.0

+ `BigIntJs`: JS implementation of `BigInt`

### Breaking changes

+ `BigInt`: `BigInteger`
+ `BigIntVm`: `BigIntegerVm`
+ `BigIntRef`: `BigIntegerRef`

## 1.0.0

+ `BigInteger` interface desing
+ `BigIntegerVM`: VM implementation of `BigInteger`
+ `BigIntegerRef` for writing efficient equations
+ `EfficientBigMath` for writing efficient equations


test/vm/assign.dart
import 'dart:math' as math;
import 'package:fatint/fatint.dart';
import 'package:test/test.dart';

void main() {
  group('BitManipulation', () {
    test('SetBit', () {
      final five = new BigInt()..assignInt = 5;
      expect(five.value, 5);
    });
  });
}


test/common/math.dart
import 'dart:math' as math;
import 'package:fatint/fatint.dart';
import 'package:test/test.dart';

void main() {
  group('math', () {
    test('mod', () {
      expect((bigInt(25) % bigInt(7)).toString(), '4');
      expect((bigHex('abcd1234') % bigHex('beef')).toString(radix: 16), 'b60c');
    });

    test('pow', () {
      expect((bigInt(3).pow(3)).toString(radix: 16), '1b');
    });
  });
}


test/vm/construction.dart
import 'dart:math' as math;
import 'package:fatint/fatint.dart';
import 'package:test/test.dart';

void main() {
  group('VM.Constructor', () {
    test('Default', () {
      expect(new BigInt().value, null);
      expect(new BigInt(0).value, 0);

      expect(new BigInt(10).toString(), "10");

      expect(new BigInt(math.pow(2, 64)).value.toString(),
          '18446744073709551616');
      expect(new BigInt(math.pow(2, 64)).bitLength, 65);

      expect(new BigInt(math.pow(2, 100)).value.toString(),
          '1267650600228229401496703205376');
      expect(new BigInt(math.pow(2, 100)).bitLength, 101);

      expect(new BigInt(-math.pow(2, 64)).value.toString(),
          '-18446744073709551616');
      expect(new BigInt(-math.pow(2, 64)).bitLength, 64);

      expect(
          new BigInt(
                  0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
              .toString(radix: 16),
          "fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
    });

    test('fromNum', () {
      expect(new BigInt.fromNum(0).value, 0);
      expect(new BigInt.fromNum(25.0).value, 25);
      expect(new BigInt.fromNum(2e64).toString(),
          '20000000000000000426408380189087937446025157425359298935486676992');
      expect(new BigInt.fromNum(2e64).bitLength, 214);
      expect(new BigInt.fromNum(-2e64).value.toString(),
          '-20000000000000000426408380189087937446025157425359298935486676992');
      expect(new BigInt.fromNum(-2e64).bitLength, 214);
    });

    test('fromBytes', () {
      expect(new BigInt.fromBytes([0x1]).toString(), '1');
      expect(new BigInt.fromBytes([0x1, 0]).toString(radix: 16), '100');
      expect(
          new BigInt.fromBytes([0xAA, 0, 0]).toString(radix: 16), 'aa0000');
      expect(new BigInt.fromBytes([0xFF, 0, 0, 0]).toString(radix: 16),
          'ff000000');
    });

    test('fromString.hex', () {
      expect(new BigInt.fromString('abcd1234', 16).toString(radix: 16),
          'abcd1234');
      expect(new BigInt.fromString('beef', 16).toString(radix: 16), 'beef');
    });
  });
}


lib/src/js/js.dart
library fatint.js;

import 'dart:typed_data';
import 'package:biginteger_js/biginteger_js.dart';

import '../bignum.dart';

part 'ref.dart';

/// Emulates signed and unsigned big integer on platforms where [int]
/// supports big int.
class BigIntJs implements BigInt {
  static BigIntJs get zero => new BigIntJs(0);

  static BigIntJs get one => new BigIntJs(1);

  /// The actual value
  BigInteger _value;

  BigIntJs._(BigInteger data) : _value = MakeBigInt.fromAnother(data);

  /// Creates a [BigIntJs] with integer value [value]
  ///
  /// Example:
  ///
  ///     final five = new BigIntegerJs(5);
  BigIntJs([int value = 0]) : _value = MakeBigInt.fromNum(value);

  /// Creates a [BigIntJs] with integer value obtained by converting [data]
  /// to integer.
  ///
  /// Example:
  ///
  ///     final five = new BigIntegerJs.fromNum(5.0);
  BigIntJs.fromNum(num data) : _value = MakeBigInt.fromNum(data.toInt());

  BigIntJs.from8(int a, [int b, int c, int d, int e, int f, int g, int h]) {
    final data = new List<int>();
    data.add(a & 0xFF);

    for (int v in <int>[b, c, d, e, f, g, h]) {
      if (v == null) break;

      data.add(v & 0xFF);
    }
    _value = MakeBigInt.fromBytes(data);
  }

  BigIntJs.from16(int a, [int b, int c, int d]) {
    final data = new List<int>();
    data.add((a >> 8) & 0xFF);
    data.add(a & 0xFF);

    for (int v in <int>[b, c, d]) {
      if (v == null) break;

      data.add((v >> 8) & 0xFF);
      data.add(v & 0xFF);
    }
    _value = MakeBigInt.fromBytes(data);
  }

  BigIntJs.from32(int a, [int b]) {
    _value = MakeBigInt.fromNum(a & 0xFFFFFFFF);
    if (b != null) {
      _value.shiftLeft(32).or(b & 0xFFFFFFFF);
    }
  }

  /// Creates a [BigIntJs] from [bytes]
  ///
  /// Example:
  ///
  ///     final five = new BigIntegerJs.fromBytes([0x5]);
  BigIntJs.fromBytes(List<int> bytes, [bool negate = false]) {
    assignBytes = bytes;
    if(negate) _value = _value.negate();
  }

  /// Creates a [BigIntJs] by parsing integer representation of [dataStr].
  ///
  /// [radix] can be used to parse integers encoded with radix other
  /// than 10. [radix] defaults to 10.
  ///
  /// Example:
  ///     final five = new BigIntegerJs.fromString('5');
  ///     final beef = new BigIntegerJs.fromString('beef', 16);
  BigIntJs.fromString(String dataStr, [int radix]) {
    setString(dataStr, radix);
  }

  /// Creates a [BigIntJs] from [sign] and [bytes].
  ///
  /// Example:
  ///
  ///     final five = new BigIntegerJs.fromSignedBytes([0xFB]);
  factory BigIntJs.fromSignedBytes(List<int> magnitude) {
    final self = new BigIntJs();
    self.assignSignedBytes = magnitude;
    return self;
  }

  /// Returns new [BigInt] with same value as [this].
  BigIntJs get clone => new BigIntJs._(this._value);

  /// Returns a [BigIntRef] of clone of [this].
  ///
  /// Example:
  ///     final other = new BigIntegerJs(5);
  ///     final five = new BigIntegerJs(20);
  ///     five.assign += other; // five.toString() == '25'
  BigIntJs get assign => new BigIntRefJs(this);

  /// Assigns value of [this] to the value of [other]
  ///
  /// Example:
  ///     final other = new BigIntegerJs(5);
  ///     final five = new BigIntegerJs();
  ///     five.assign = other;  // five.toString() = '5'
  set assign(covariant BigIntJs other) => _value = other._value;

  /// Sets integer value to [value]
  ///
  /// Example:
  ///     final five = new BigIntegerJs();
  ///     five.assignInt = 5;
  set assignInt(int v) => _value = MakeBigInt.fromNum(v);

  set assignBin(String v) => _value = MakeBigInt.fromString(v, 2);

  set assignOct(String v) => _value = MakeBigInt.fromString(v, 8);

  set assignDec(String v) => _value = MakeBigInt.fromString(v, 10);

  set assignHex(String v) => _value = MakeBigInt.fromString(v, 16);

  set assignBytes(List<int> bytes) {
    _value = MakeBigInt.fromBytes(bytes);
  }

  set assignSignedBytes(List<int> bytes) {
    if (bytes.length == 0) {
      _value = MakeBigInt.fromNum();
      return;
    }

    final bool isNeg = (bytes[0] & 0x80) != 0;

    final ubytes = new Uint8List(bytes.length);

    for (int i = 0; i < bytes.length; i++) {
      final int ob = bytes[i];

      if(!isNeg) {
        ubytes[i] = ob & 0xFF;
      } else {
        ubytes[i] = (~ob) & 0xFF;
      }
    }

    _value = MakeBigInt.fromBytes(ubytes);
    _value = _value.next().negate();
  }

  /// Parses the integer value from its string representation [dataStr].
  ///
  /// [radix] can be used to parse integers encoded with radix other
  /// than 10. [radix] defaults to 10.
  ///
  /// Example:
  ///     final five = new BigIntegerJs();
  ///     five.setString('5');
  ///     final beef = new BigIntegerJs();
  ///     five.setString('beef', 16);
  void setString(String dataStr, [int radix]) {
    _value = MakeBigInt.fromString(dataStr, radix);
  }

  /// Return string representation with [radix].
  String toString({int radix = 10}) {
    return _value.toString(radix);
  }

  List<int> get toBytes {
    final ret = new Uint8List(byteLength);

    for(int i = 0; i < byteLength; i++) {
      ret[i] = _value.shiftRight(byteLength * 8).and(0xFF).toJSNumber();
    }

    return ret;
  }

  /// Return [this] + [other]. Does not modify either [this] or [other]!
  BigIntJs operator +(covariant BigIntJs other) =>
      new BigIntJs._(_value.add(other._value));

  /// Return [this] - [other]. Does not modify either [this] or [other]!
  BigIntJs operator -(covariant BigIntJs other) =>
      new BigIntJs._(_value.subtract(other._value));

  /// Return [this] * [other]. Does not modify either [this] or [other]!
  BigIntJs operator *(covariant BigIntJs other) =>
      new BigIntJs._(_value.multiply(other._value));

  /// Return [this] / [other]. Does not modify either [this] or [other]!
  BigIntJs operator /(covariant BigIntJs other) =>
      new BigIntJs._(_value.divide(other._value));

  /// Return [this] ~/ [other]. Does not modify either [this] or [other]!
  BigIntJs operator ~/(covariant BigIntJs other) =>
      new BigIntJs._(_value.divide(other._value));

  /// Return [this] % [other]. Does not modify either [this] or [other]!
  BigIntJs operator %(covariant BigIntJs other) =>
      new BigIntJs._(_value.mod(other._value));

  /// Return -[this]. Does not modify [this]!
  BigIntJs operator -() => new BigIntJs._(_value.negate());

  /// Returns [this] & [other].Does not modify either [this] or [other]!
  BigIntJs operator &(covariant BigIntJs other) =>
      new BigIntJs._(_value.and(other._value));

  /// Returns [this] | [other].Does not modify either [this] or [other]!
  BigIntJs operator |(covariant BigIntJs other) =>
      new BigIntJs._(_value.or(other._value));

  /// Returns [this] ^ [other].Does not modify either [this] or [other]!
  BigIntJs operator ^(covariant BigIntJs other) =>
      new BigIntJs._(_value.xor(other._value));

  /// Returns ~[this].Does not modify [this]!
  BigIntJs operator ~() => new BigIntJs._(_value.not());

  BigIntJs andNot(covariant BigIntJs other) =>
      new BigIntJs._(_value.and(other._value.not()));

  /// Return positive integer if [this] > [other], negative integer if [this] < [other],
  /// 0 if equal.
  int compareTo(covariant BigIntJs other) => _value.compareTo(other._value);

  bool equals(covariant BigIntJs other) =>
      this.compareTo(other) == 0 ? true : false;

  bool operator <(covariant BigIntJs other) =>
      compareTo(other) < 0 ? true : false;

  bool operator <=(covariant BigIntJs other) =>
      compareTo(other) <= 0 ? true : false;

  bool operator >(covariant BigIntJs other) =>
      compareTo(other) > 0 ? true : false;

  bool operator >=(covariant BigIntJs other) =>
      compareTo(other) >= 0 ? true : false;

  bool operator ==(other) {
    if (other is! BigIntJs) return false;
    return compareTo(other) == 0 ? true : false;
  }

  /// Returns value left shifted by [n] bits.  Does not modify
  /// [this]!
  BigIntJs operator <<(int n) => new BigIntJs._(_value.shiftLeft(n));

  /// Returns value right shifted by [n] bits.  Does not modify
  /// [this]!
  BigIntJs operator >>(int n) => new BigIntJs._(_value.shiftRight(n));

  /// Returns absolute value
  BigIntJs abs() => new BigIntJs._(_value.abs());

  DivResult divide(BigInt denominator) {
    return new DivResult(q: this / denominator, r: remainder(denominator));
  }

  /// Returns [this] % [denominator].
  BigIntJs remainder(covariant BigIntJs denominator) =>
      new BigIntJs._(_value.remainder(denominator._value));

  /// Returns [this] ^ [e]
  BigIntJs pow(int e) {
    if (e > 0xffffffff || e < 1) return BigIntJs.one;
    return new BigIntJs._(_value.pow(e));
  }

  /// Returns [this]^[e] % [m]
  ///
  /// (HAC 14.85)
  BigIntJs modPow(covariant BigIntJs e, covariant BigIntJs m) =>
      new BigIntJs._(_value.modPow(e._value, m._value));

  /// Returns 1/[this] % m
  ///
  /// (HAC 14.61)
  BigIntJs modInverse(covariant BigIntJs m) =>
      new BigIntJs._(_value.modInv(m._value));

  BigIntJs min(covariant BigIntJs a) => (this.compareTo(a) < 0) ? this : a;

  BigIntJs max(covariant BigIntJs a) => (this.compareTo(a) > 0) ? this : a;

  /// Returns gcd([this], [other])
  ///
  /// (HAC 14.54)
  BigIntJs gcd(covariant BigIntJs other) =>
      new BigIntJs._(MakeBigInt.gcd(_value, other._value));

  /// Returns [this] | (1 << [bitPos]). Does not modify [this].
  BigIntJs setBit(int bitPos) =>
      new BigIntJs._(_value.or(MakeBigInt.one.shiftLeft(bitPos)));

  /// Returns [this] & ~(1 << [bitPos]). Does not modify [this].
  BigIntJs clearBit(int bitPos) =>
      new BigIntJs._(_value.and(MakeBigInt.one.shiftLeft(bitPos).not()));

  /// Returns [this] ^ (1 << [bitPos]). Does not modify [this].
  BigIntJs toggleBit(int bitPos) =>
      new BigIntJs._(_value.xor(MakeBigInt.one.shiftLeft(bitPos)));

  /// Returns [this] % [n] when n < 2^26
  BigIntJs modInt(int n) => new BigIntJs._(_value.mod(n));

  /// Returns [this]^[e] % [m] when 0 <= e < 2^32.
  BigIntJs modPowInt(int e, covariant BigIntJs m) =>
      new BigIntJs._(_value.modPow(e, m._value));

  /// Returns true if value is even
  bool get isEven => _value.isEven();

  /// Returns true if value is odd
  bool get isOdd => _value.isOdd();

  /// Returns number of bits required to represent [this].
  int get bitLength {
    if(_value.isZero()) return 0;

    BigInteger data = _value;
    int ret = 0;
    while(!data.isZero()) {
      ret++;
      data = data.shiftRight(1);
    }
    return ret;
  }

  /// Returns number of bytes required to represent [this].
  int get byteLength => (bitLength + 7) ~/ 8;

  /// Returns value as [int].
  int get value => _value.toJSNumber();

  /// Returns value as 8-bit integer.
  int get value8 => value & 0xff;

  /// Returns value as 16-bit integer.
  int get value16 => value & 0xffff;

  /// Returns value as 32-bit integer.
  int get value32 => value & 0xffffffff;

  /// Returns value as 53-bit integer.
  int get value53 => value & 0x1fffffffffffff;

  /// Returns true iff bit at position [bitPos] is set.
  bool testBit(int bitPos) =>
      !_value.and(MakeBigInt.one.shiftLeft(bitPos)).isZero();

  int get sign => _value.isZero() ? 0 : _value.isNegative() ? -1 : 1;

  bool get isNegative => _value.isNegative();
}


lib/src/vm/ref.dart
part of fatint.dartvm;

class BigIntRefVm extends BigIntVm implements BigIntRef {
  final BigIntVm deref;

  BigIntRefVm get ref => this;

  int get _value => deref._value;

  set _value(int v) => deref._value = v;

  BigIntRefVm(this.deref);

  BigIntRefVm operator +(covariant BigIntVm other) {
    _value += other._value;
    return this;
  }

  BigIntRefVm operator -(covariant BigIntVm other) {
    _value -= other._value;
    return this;
  }

  BigIntRefVm operator *(covariant BigIntVm other) {
    _value *= other._value;
    return this;
  }

  BigIntRefVm operator /(covariant BigIntVm other) {
    _value ~/= other._value;
    return this;
  }

  /// Return [this] ~/ [other].
  BigIntRefVm operator ~/(covariant BigIntVm other) {
    _value ~/= other._value;
    return this;
  }

  /// Return [this] % [other].
  BigIntRefVm operator %(covariant BigIntVm other) {
    _value %= other._value;
    return this;
  }

  BigIntRefVm operator -() {
    _value = -_value;
    return this;
  }

  BigIntRefVm operator &(covariant BigIntVm other) {
    _value &= other._value;
    return this;
  }

  BigIntRefVm operator |(covariant BigIntVm other) {
    _value |= other._value;
    return this;
  }

  BigIntRefVm operator ^(covariant BigIntVm other) {
    _value ^= other._value;
    return this;
  }

  BigIntRefVm operator ~() {
    _value = ~_value;
    return this;
  }

  BigIntRefVm andNot(covariant BigIntVm other) {
    _value &= ~other._value;
    return this;
  }

  BigIntRefVm operator <<(int n) {
    _value <<= n;
    return this;
  }

  BigIntRefVm operator >>(int n) {
    _value >>= n;
    return this;
  }

  BigIntRefVm abs() {
    _value = _value.abs();
    return this;
  }

  BigIntRefVm remainder(covariant BigIntVm denominator) {
    _value = _value.remainder(denominator._value);
    return this;
  }

  BigIntRefVm pow(int e) {
    if (e > 0xffffffff || e < 1) return BigIntVm.one;
    _value = math.pow(_value, e);
    return this;
  }

  BigIntRefVm modPow(covariant BigIntVm e, covariant BigIntVm m) {
    _value = _value.modPow(e._value, m._value);
    return this;
  }

  BigIntRefVm modInverse(covariant BigIntVm m) {
    _value = _value.modInverse(m._value);
    return this;
  }

  BigIntRefVm min(covariant BigIntVm other) {
    _value = math.min(_value, other._value);
    return this;
  }

  BigIntRefVm max(covariant BigIntVm other) {
    _value = math.max(_value, other._value);
    return this;
  }

  BigIntRefVm gcd(covariant BigIntVm other) {
    _value = _value.gcd(other._value);
    return this;
  }

  /// Returns [this] | (1 << [bitPos]).
  BigIntRefVm setBit(int bitPos) {
    _value |= 1 << bitPos;
    return this;
  }

  /// Returns [this] & ~(1 << [bitPos]).
  BigIntRefVm clearBit(int bitPos) {
    _value &= ~(1 << bitPos);
    return this;
  }

  /// Returns [this] ^ (1 << [bitPos]).
  BigIntRefVm toggleBit(int bitPos) {
    _value ^= 1 << bitPos;
    return this;
  }

  /// Returns [this] % [n] when n < 2^26
  BigIntRefVm modInt(int n) {
    _value %= n;
    return this;
  }

  /// Returns [this]^[e] % [m] when 0 <= e < 2^32.
  BigIntRefVm modPowInt(int e, covariant BigIntVm m) {
    _value = _value.modPow(e, m._value);
    return this;
  }
}


lib/src/vm/vm.dart
library fatint.dartvm;

import 'dart:math' as math;
import 'dart:typed_data';

import '../bignum.dart';

part 'ref.dart';

/// Emulates signed and unsigned big integer on platforms where [int]
/// supports big int.
class BigIntVm implements BigInt {
  static BigIntVm get zero => new BigIntVm(0);

  static BigIntVm get one => new BigIntVm(1);

  /// The actual value
  int _value;

  /// Creates a [BigIntVm] with integer value [_value]
  ///
  /// Example:
  ///
  ///     final five = new BigIntVm(5);
  BigIntVm([this._value = 0]);

  /// Creates a [BigIntVm] with integer value obtained by converting [data]
  /// to integer.
  ///
  /// Example:
  ///
  ///     final five = new BigIntVm.fromNum(5.0);
  BigIntVm.fromNum(num data) : _value = data.toInt();

  BigIntVm.from8(int a, [int b, int c, int d, int e, int f, int g, int h]) {
    _value = a & 0xFF;

    for (int v in <int>[b, c, d, e, f, g, h]) {
      if (v == null) break;

      _value = (_value << 8) | (v & 0xFF);
    }
  }

  BigIntVm.from16(int a, [int b, int c, int d]) {
    _value = a & 0xFFFF;

    for (int v in <int>[b, c, d]) {
      if (v == null) break;

      _value = (_value << 16) | (v & 0xFFFF);
    }
  }

  BigIntVm.from32(int a, [int b]) {
    _value = a & 0xFFFFFFFF;
    if (b != null) _value = (_value << 32) | (b & 0xFFFFFFFF);
  }

  /// Creates a [BigIntVm] from [bytes]
  ///
  /// Example:
  ///
  ///     final five = new BigIntVm.fromBytes([0x5]);
  BigIntVm.fromBytes(List<int> bytes, [bool negate = false]) {
    assignBytes = bytes;
    if(negate) _value = -_value;
  }

  /// Creates a [BigIntVm] by parsing integer representation of [dataStr].
  ///
  /// [radix] can be used to parse integers encoded with radix other
  /// than 10. [radix] defaults to 10.
  ///
  /// Example:
  ///     final five = new BigIntVm.fromString('5');
  ///     final beef = new BigIntVm.fromString('beef', 16);
  BigIntVm.fromString(String dataStr, [int radix]) {
    setString(dataStr, radix);
  }

  /// Creates a [BigIntVm] from [signum] and [bytes].
  ///
  /// Example:
  ///
  ///     final five = new BigIntVm.fromSignedBytes([0x5]);
  factory BigIntVm.fromSignedBytes(List<int> bytes) {
    final self = new BigIntVm();
    self.assignSignedBytes = bytes;
    return self;
  }

  /// Returns new [BigInt] with same value as [this].
  BigIntVm get clone => new BigIntVm(this._value);

  /// Returns a [BigIntRef] of clone of [this].
  ///
  /// Example:
  ///     final other = new BigIntVm(5);
  ///     final five = new BigIntVm(20);
  ///     five.assign += other; // five.toString() == '25'
  BigIntVm get assign => new BigIntRefVm(clone);

  /// Assigns value of [this] to the value of [other]
  ///
  /// Example:
  ///     final other = new BigIntVm(5);
  ///     final five = new BigIntVm();
  ///     five.assign = other;  // five.toString() = '5'
  set assign(covariant BigIntVm other) => _value = other._value;

  /// Sets integer value to [value]
  ///
  /// Example:
  ///     final five = new BigIntVm();
  ///     five.assignInt = 5;
  set assignInt(int v) => _value = v;

  set assignBin(String v) => _value = int.parse(v, radix: 2, onError: (_) => 0);

  set assignOct(String v) => _value = int.parse(v, radix: 8, onError: (_) => 0);

  set assignDec(String v) =>
      _value = int.parse(v, radix: 10, onError: (_) => 0);

  set assignHex(String v) =>
      _value = int.parse(v, radix: 16, onError: (_) => 0);

  set assignBytes(List<int> bytes) {
    _value = 0;

    for (int v in bytes ?? <int>[]) {
      _value = (_value << 8) | (v & 0xFF);
    }
  }

  set assignSignedBytes(List<int> bytes) {
    if (bytes.length == 0) {
      _value = 0;
      return;
    }

    final bool isNeg = (bytes[0] & 0x80) != 0;

    _value = 0;
    for (int byte in bytes) {
      _value <<= 8;
      if (!isNeg)
        _value |= byte & 0xFF;
      else
        _value |= (~byte) & 0xFF;
    }

    if (isNeg) {
      _value++;
      _value = -_value;
    }
  }

  /// Parses the integer value from its string representation [dataStr].
  ///
  /// [radix] can be used to parse integers encoded with radix other
  /// than 10. [radix] defaults to 10.
  ///
  /// Example:
  ///     final five = new BigIntVm();
  ///     five.setString('5');
  ///     final beef = new BigIntVm();
  ///     five.setString('beef', 16);
  void setString(String dataStr, [int radix]) {
    _value = int.parse(dataStr, radix: radix, onError: (_) => 0);
  }

  /// Return string representation with [radix].
  String toString({int radix = 10}) {
    return _value.toRadixString(radix);
  }

  List<int> get toBytes {
    final ret = new Uint8List(byteLength);

    for (int i = 0; i < ret.length; i++) {
      ret[i] = (_value >> ((ret.length - i - 1) * 8)) & 0xFF;
    }

    return ret;
  }

  /// Return [this] + [other]. Does not modify either [this] or [other]!
  BigIntVm operator +(covariant BigIntVm other) =>
      new BigIntVm(_value + other._value);

  /// Return [this] - [other]. Does not modify either [this] or [other]!
  BigIntVm operator -(covariant BigIntVm other) =>
      new BigIntVm(_value - other._value);

  /// Return [this] * [other]. Does not modify either [this] or [other]!
  BigIntVm operator *(covariant BigIntVm other) =>
      new BigIntVm(_value * other._value);

  /// Return [this] / [other]. Does not modify either [this] or [other]!
  BigIntVm operator /(covariant BigIntVm other) =>
      new BigIntVm(_value ~/ other._value);

  /// Return [this] ~/ [other]. Does not modify either [this] or [other]!
  BigIntVm operator ~/(covariant BigIntVm other) =>
      new BigIntVm(_value ~/ other._value);

  /// Return [this] % [other]. Does not modify either [this] or [other]!
  BigIntVm operator %(covariant BigIntVm other) =>
      new BigIntVm(_value % other._value);

  /// Return -[this]. Does not modify [this]!
  BigIntVm operator -() => new BigIntVm(-_value);

  /// Returns [this] & [other].Does not modify either [this] or [other]!
  BigIntVm operator &(covariant BigIntVm other) =>
      new BigIntVm(_value & other._value);

  /// Returns [this] | [other].Does not modify either [this] or [other]!
  BigIntVm operator |(covariant BigIntVm other) =>
      new BigIntVm(_value | other._value);

  /// Returns [this] ^ [other].Does not modify either [this] or [other]!
  BigIntVm operator ^(covariant BigIntVm other) =>
      new BigIntVm(_value ^ other._value);

  /// Returns ~[this].Does not modify [this]!
  BigIntVm operator ~() => new BigIntVm(~_value);

  BigIntVm andNot(covariant BigIntVm other) =>
      new BigIntVm(_value & ~other._value);

  /// Return positive integer if [this] > [other], negative integer if [this] < [other],
  /// 0 if equal.
  int compareTo(covariant BigIntVm other) => _value.compareTo(other._value);

  bool equals(covariant BigIntVm other) =>
      this.compareTo(other) == 0 ? true : false;

  bool operator <(covariant BigIntVm other) =>
      compareTo(other) < 0 ? true : false;

  bool operator <=(covariant BigIntVm other) =>
      compareTo(other) <= 0 ? true : false;

  bool operator >(covariant BigIntVm other) =>
      compareTo(other) > 0 ? true : false;

  bool operator >=(covariant BigIntVm other) =>
      compareTo(other) >= 0 ? true : false;

  bool operator ==(other) {
    if (other is! BigIntVm) return false;
    return compareTo(other) == 0 ? true : false;
  }

  /// Returns value left shifted by [n] bits.  Does not modify
  /// [this]!
  BigIntVm operator <<(int n) => new BigIntVm(_value << n);

  /// Returns value right shifted by [n] bits.  Does not modify
  /// [this]!
  BigIntVm operator >>(int n) => new BigIntVm(_value >> n);

  /// Returns absolute value
  BigIntVm abs() => new BigIntVm(_value.abs());

  DivResult divide(BigInt denominator) {
    return new DivResult(q: this / denominator, r: remainder(denominator));
  }

  /// Returns [this] % [denominator].
  BigIntVm remainder(covariant BigIntVm denominator) =>
      new BigIntVm(_value.remainder(denominator._value));

  /// Returns [this] ^ [e]
  BigIntVm pow(int e) {
    if (e > 0xffffffff || e < 1) return BigIntVm.one;
    return new BigIntVm(math.pow(_value, e));
  }

  /// Returns [this]^[e] % [m]
  ///
  /// (HAC 14.85)
  BigIntVm modPow(covariant BigIntVm e, covariant BigIntVm m) =>
      new BigIntVm(_value.modPow(e._value, m._value));

  /// Returns 1/[this] % m
  ///
  /// (HAC 14.61)
  BigIntVm modInverse(covariant BigIntVm m) =>
      new BigIntVm(_value.modInverse(m._value));

  BigIntVm min(covariant BigIntVm a) => (this.compareTo(a) < 0) ? this : a;

  BigIntVm max(covariant BigIntVm a) => (this.compareTo(a) > 0) ? this : a;

  /// Returns gcd([this], [other])
  ///
  /// (HAC 14.54)
  BigIntVm gcd(covariant BigIntVm other) =>
      new BigIntVm(_value.gcd(other._value));

  /// Returns [this] | (1 << [bitPos]). Does not modify [this].
  BigIntVm setBit(int bitPos) => new BigIntVm(_value | (1 << bitPos));

  /// Returns [this] & ~(1 << [bitPos]). Does not modify [this].
  BigIntVm clearBit(int bitPos) => new BigIntVm(_value & ~(1 << bitPos));

  /// Returns [this] ^ (1 << [bitPos]). Does not modify [this].
  BigIntVm toggleBit(int bitPos) => new BigIntVm(_value ^ (1 << bitPos));

  /// Returns [this] % [n] when n < 2^26
  BigIntVm modInt(int n) => new BigIntVm(_value % n);

  /// Returns [this]^[e] % [m] when 0 <= e < 2^32.
  BigIntVm modPowInt(int e, covariant BigIntVm m) =>
      new BigIntVm(_value.modPow(e, m._value));

  /// Returns true if value is even
  bool get isEven => _value.isEven;

  /// Returns true if value is odd
  bool get isOdd => _value.isOdd;

  /// Returns number of bits required to represent [this].
  int get bitLength => _value.bitLength;

  /// Returns number of bytes required to represent [this].
  int get byteLength => (bitLength + 7) ~/ 8;

  /// Returns value as [int].
  int get value => _value;

  /// Returns value as 8-bit integer.
  int get value8 => _value & 0xff;

  /// Returns value as 16-bit integer.
  int get value16 => _value & 0xffff;

  /// Returns value as 32-bit integer.
  int get value32 => _value & 0xffffffff;

  /// Returns value as 53-bit integer.
  int get value53 => _value & 0x1fffffffffffff;

  /// Returns true iff bit at position [bitPos] is set.
  bool testBit(int bitPos) => (_value & (1 << bitPos)) != 0;

  int get sign => _value.sign;

  bool get isNegative => _value.isNegative;
}
